expr/expr.go

```diff
@@ -3266,6 +3267,82 @@ func EvalExpr(e *expr, from, until int32, values map[MetricRequest][]*MetricData
 		}
 
 		return results, nil
+
+	case "cactiStyle": // cactiStyle(seriesList, system=None, units=None)
+		// Get the series data
+		original, err := getSeriesArg(e.args[0], from, until, values)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get the arguments
+		system, err := getStringNamedOrPosArgDefault(e, "system", 1, "")
+		if err != nil {
+			return nil, err
+		}
+		unit, err := getStringNamedOrPosArgDefault(e, "units", 2, "")
+		if err != nil {
+			return nil, err
+		}
+
+		// Deal with each of the series
+		for _, a := range original {
+			// Calculate min, max, current
+			//
+			// This saves calling summarizeValues 3 times and looping over
+			// the metrics 3 times
+			//
+			// For min:
+			// Ignoring any absent values and inf (if we have a value)
+			// Using summarizeValues("min", ...) results in incorrect values, when absent
+			// values are present
+			//
+			val_min := math.Inf(-1)
+			val_current := math.Inf(-1)
+			val_max := math.Inf(-1)
+			for i, av := range a.GetValues() {
+				if !a.IsAbsent[i] && (val_min == math.Inf(-1) || av < val_min) {
+					val_min = av
+				}
+				if !a.IsAbsent[i] && av > val_max {
+					val_max = av
+				}
+				val_current = av
```

Comment:
Yep - also has an issue with the tests failing due to changed data
`	expr_test.go:2305: cactiStyle(): source data was modified key {metric1 0 1} index 0 want:
		name:"metric1" startTime:1475007879 stopTime:1475007889 stepTime:1 values:0 values:20531.733333333334 values:20196.4 values:17925.333333333332 values:20950.4 values:35168.13333333333 values:19965.866666666665 values:24556.4 values:22266.4 values:58039.86666666667 isAbsent:true isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false 
		 got:
		name:"metric1 Current: 58k Max: 58k Min: 18k" startTime:1475007879 stopTime:1475007889 stepTime:1 values:0 values:20531.733333333334 values:20196.4 values:17925.333333333332 values:20950.4 values:35168.13333333333 values:19965.866666666665 values:24556.4 values:22266.4 values:58039.86666666667 isAbsent:true isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false isAbsent:false 
	expr_test.go:2320: bad name for cactiStyle() metric 0: got metric1 Current: 58k Max: 58k Min: 18k, want metric1 Current:58k Max:58k Min:18k`

Changed to copying the values into a new array and returning that, which seems to satisfy the tests